packages/node/src/integrations/tracing/vercelai/instrumentation.ts

```diff
@@ -35,6 +40,101 @@ interface RecordingOptions {
   recordOutputs?: boolean;
 }
 
+interface ToolError {
+  type: 'tool-error' | 'tool-result' | 'tool-call';
+  toolCallId: string;
+  toolName: string;
+  input?: {
+    [key: string]: unknown;
+  };
+  error: Error;
+  dynamic?: boolean;
+}
+
+function isToolError(obj: unknown): obj is ToolError {
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+
+  const candidate = obj as Record<string, unknown>;
+  return (
+    'type' in candidate &&
+    'error' in candidate &&
+    'toolName' in candidate &&
+    'toolCallId' in candidate &&
+    candidate.type === 'tool-error' &&
+    candidate.error instanceof Error
+  );
+}
+
+/**
+ * Check for tool errors in the result and capture them
+ * Tool errors are not rejected in Vercel V5, it is added as metadata to the result content
+ */
+function checkResultForToolErrors(result: unknown | Promise<unknown>): void {
+  if (typeof result !== 'object' || result === null || !('content' in result)) {
+    return;
+  }
+
+  const resultObj = result as { content: Array<object> };
+  if (!Array.isArray(resultObj.content)) {
+    return;
+  }
+
+  for (const item of resultObj.content) {
+    if (isToolError(item)) {
+      // Try to get the span associated with this tool call ID
+      const associatedSpan = getSpanForToolCallId(item.toolCallId) as Span;
+
+      if (associatedSpan) {
+        // We have the span, so link the error using span and trace IDs from the span
+        const spanContext = associatedSpan.spanContext();
+
+        withScope(scope => {
+          // Set the span and trace context for proper linking
+          scope.setContext('trace', {
+            trace_id: spanContext.traceId,
+            span_id: spanContext.spanId,
+          });
+
+          // Add tool-specific tags
+          scope.setTag('vercel.ai.tool.name', item.toolName);
+          scope.setTag('vercel.ai.tool.callId', item.toolCallId);
+          scope.setTag('vercel.ai.tool.span.id', spanContext.spanId);
+          scope.setTag('vercel.ai.tool.trace.id', spanContext.traceId);
```

Comment:
I think we could use `span_id` and `trace_id` here. And is there a specific reason we use tags on the scope here instead of attributes on the span? Just wondering 🤔 


```suggestion
          scope.setTag('vercel.ai.tool.spanId', spanContext.span_id);
          scope.setTag('vercel.ai.tool.traceId', spanContext.trace_id);
```

Reply:
Good catch on the duplicate span/trace id tags — just removed them, thanks. FYI SpanContextData only has traceId and spanId. And i'm adding some tags because i'm trying to enrich the error, the span’s already ended by time we get here.